Extensions: Zoninator - Implement state for zoninator's zone locks

ice9js opened the pull request:
This PR implements the state needed to implement zone locks in the Zoninator extension.

About the zone locks themselves, Zoninator has a feature that prevents multiple people from making editing the same zone at the same time. It does that by pinging a specific endpoint on a regular interval while the user is on the edit page.  
In order to prevent the user from completely locking out everyone else forever, there is a maximum lock time for a user after which we should stop making new requests.

In its current shape, the state should allow for tracking the current expiration time as well as making sure it doesn't exceed the maximum lock period.  
Additionally it should account for a zone possibly being blocked by another user.

# Testing

Run the module's unit tests:
`npm run test-client client/extensions/zoninator/state/locks`

alexsanford commented on client/extensions/zoninator/state/locks/actions.js:
```diff
@@ -0,0 +1,59 @@
+ *
+ * @param  {Number}  siteId  Site ID
+ * @param  {Number}  zoneId  Zone ID
+ * @param  {Boolean} reset   Set to true to reset a lock
```
Is this a typo? Or can you actually "reset" a lock when you're requesting it? If that is the case, then I don't see any handling of this in the reducer.

ice9js replied:
Well, the idea was to pass that value to the following action on success in case the timer was to be reset, but I see that there's no point in doing that as I can just change the reducer to work with `ZONINATOR_REQUEST_LOCK`.

alexsanford commented on client/extensions/zoninator/state/locks/test/reducer.js:
```diff
@@ -0,0 +1,183 @@
+		} );
+	} );
+
+	describe( 'items()', () => {
```
For `items()` and `maxLockPeriod()`, should we have tests for setting these values?

ockham commented on client/extensions/zoninator/state/locks/actions.js:
```diff
@@ -0,0 +1,59 @@
+ * @param  {Date}    expires         Expiration date
+ * @param  {Number}  maxLockPeriod   Maximum number of seconds to extend the lock to
+ * @param  {Boolean} reset           Set to true to reset a lock
+ * @return {Action}                  Action object
```
I guess to be JSDoc compliant, the return type needs to be `{Object}`, no?

ockham commented on client/extensions/zoninator/state/locks/actions.js:
```diff
@@ -0,0 +1,59 @@
+ * @param  {Boolean} reset           Set to true to reset a lock
+ * @return {Action}                  Action object
+ */
+export const updateLock = ( siteId, zoneId, expires, maxLockPeriod, reset ) => ( {
```
5 args is quite a few. What about `( siteId, zoneId, { expires, maxLockPeriod, reset } )`?

Alternatively, you could define a separate `resetLock` function, especially if that happens to have a different signature (e.g. does not require `expires` and/or `maxLockPeriod` (?)). Disclaimer: I don't really know how reset works.

ockham commented on client/extensions/zoninator/state/locks/actions.js:
```diff
@@ -0,0 +1,59 @@
+ * @param  {Number}  siteId  Site ID
+ * @param  {Number}  zoneId  Zone ID
+ * @param  {Boolean} reset   Set to true to reset a lock
+ * @return {Action}          Action
```
https://github.com/Automattic/wp-calypso/pull/21110/files#r160451339

ockham commented on client/extensions/zoninator/state/locks/actions.js:
```diff
@@ -0,0 +1,59 @@
+ * @param  {Number}  siteId  Site ID
+ * @param  {Number}  zoneId  Zone ID
+ * @param  {Boolean} blocked Set to true if the zone is blocked by another user
+ * @return {Action}          Action object
```
https://github.com/Automattic/wp-calypso/pull/21110/files#r160451339

ockham commented on client/extensions/zoninator/state/locks/reducer.js:
```diff
@@ -0,0 +1,67 @@
+		[ ZONINATOR_REQUEST_LOCK_ERROR ]: () => false,
+		[ ZONINATOR_UPDATE_LOCK ]: () => false,
+	}
+);
```
Note that the data layer now provides `isRequesting` information out of the box -- see #20038 for an example 😄

ockham commented on client/extensions/zoninator/state/locks/reducer.js:
```diff
@@ -0,0 +1,67 @@
+	}
+);
+
+export const blocked = keyedReducer( 'siteId', keyedReducer( 'zoneId', isBlocked ) );
```
Maybe consider inlining `isRequesting` and `isBlocked` into `requesting` and `blocked`, respectively? Using those slight name variations to denote something rather specific (basic vs keyed reducer variations) can be a bit misleading...

ockham replied:
Also, shouldn't this reducer be called `locked` instead of `blocked`? 😄

ice9js replied:
I called it `blocked` specifically so it clearly indicates the zone is being locked by another user. Using `locked` might suggest it's locked by the current user.

ockham commented on client/extensions/zoninator/state/locks/reducer.js:
```diff
@@ -0,0 +1,67 @@
+} from '../action-types';
+
+const isRequesting = createReducer(
+	{},
```
Since this reducer generally holds a bool, it should be initialized with one. Pretty safe to assume that we aren't requesting upon app startup, so `false` should be fine.

ockham commented on client/extensions/zoninator/state/locks/reducer.js:
```diff
@@ -0,0 +1,67 @@
+export const requesting = keyedReducer( 'siteId', keyedReducer( 'zoneId', isRequesting ) );
+
+const isBlocked = createReducer(
+	{},
```
https://github.com/Automattic/wp-calypso/pull/21110/files#r160455307

ockham commented on client/extensions/zoninator/state/locks/reducer.js:
```diff
@@ -0,0 +1,67 @@
+export const blocked = keyedReducer( 'siteId', keyedReducer( 'zoneId', isBlocked ) );
+
+const lockInit = createReducer(
+	{},
```
Similar to https://github.com/Automattic/wp-calypso/pull/21110/files#r160455307, I'd avoid 'empty object' as an initial value for a reducer that's supposed to hold a flat data type. Since we're dealing with an integer here, would `0` work?

ockham commented on client/extensions/zoninator/state/locks/reducer.js:
```diff
@@ -0,0 +1,67 @@
+export const created = keyedReducer( 'siteId', keyedReducer( 'zoneId', lockInit ) );
+
+const lock = createReducer(
+	{},
```
https://github.com/Automattic/wp-calypso/pull/21110/files#r160456496

ockham commented on client/extensions/zoninator/state/locks/reducer.js:
```diff
@@ -0,0 +1,67 @@
+export const items = keyedReducer( 'siteId', keyedReducer( 'zoneId', lock ) );
+
+const lockSettings = createReducer(
+	{},
```
https://github.com/Automattic/wp-calypso/pull/21110/files#r160456496

ockham commented on client/extensions/zoninator/state/locks/selectors.js:
```diff
@@ -0,0 +1,76 @@
+ * @return {Number}        Lock expiration date
+ */
+export const getLock = ( state, siteId, zoneId ) =>
+	get( getLocksState( state ), [ 'items', siteId, zoneId ], 0 );
```
Ah-hah! 😄 The fallback value here (and in selectors below) might indicate could initial values for the reducers above (in which case you might be able to drop them here).

ockham commented on client/extensions/zoninator/state/locks/reducer.js:
```diff
@@ -0,0 +1,67 @@
+const lockInit = createReducer(
+	{},
+	{
+		[ ZONINATOR_UPDATE_LOCK ]: ( state, { reset } ) => ( reset ? new Date().getTime() : state ),
```
On the issue of using timestamps in reducers, I've recently seen the following pattern #21114 to make the reducer pure (so it's completely determined by data it is fed by the action, rather than something opaque than `Date().getTime()` here). _tl;dr:_ Consider moving the `Date().getTime()` call to the corresponding action, and just pass the value to the reducer.

ockham commented on client/extensions/zoninator/state/locks/test/selectors.js:
```diff
@@ -0,0 +1,258 @@
+	};
+
+	describe( 'getLock()', () => {
+		const state = tap( cloneDeep( emptyState ), stateTree => {
```
Since this is a test, I'd suggest defining `state` explicitly (as opposed to that `tap`/`cloneDeep`/modifying function combination -- looks a bit fragile, TBH).

ockham commented on client/extensions/zoninator/state/locks/test/selectors.js:
```diff
@@ -0,0 +1,258 @@
+	} );
+
+	describe( 'isBlocked()', () => {
+		const state = tap( cloneDeep( emptyState ), stateTree => {
```
https://github.com/Automattic/wp-calypso/pull/21110/files#r160462770

ockham commented on client/extensions/zoninator/state/locks/test/selectors.js:
```diff
@@ -0,0 +1,258 @@
+	} );
+
+	describe( 'isExpired()', () => {
+		const state = tap( cloneDeep( emptyState ), stateTree => {
```
https://github.com/Automattic/wp-calypso/pull/21110/files#r160462770

ockham commented on client/extensions/zoninator/state/locks/test/selectors.js:
```diff
@@ -0,0 +1,258 @@
+		} );
+
+		test( "should return false if the expiry time hasn't passed yet", () => {
+			const unexpiredState = tap( cloneDeep( state ), stateTree => {
```
https://github.com/Automattic/wp-calypso/pull/21110/files#r160463056

ockham commented on client/extensions/zoninator/state/locks/test/selectors.js:
```diff
@@ -0,0 +1,258 @@
+		} );
+
+		test( 'should return true if the expiry time has passed', () => {
+			const expiredState = tap( cloneDeep( state ), stateTree => {
```
https://github.com/Automattic/wp-calypso/pull/21110/files#r160463056

ockham commented on client/extensions/zoninator/state/locks/test/selectors.js:
```diff
@@ -0,0 +1,258 @@
+	} );
+
+	describe( 'isRequesting()', () => {
+		const state = tap( cloneDeep( emptyState ), stateTree => {
```
https://github.com/Automattic/wp-calypso/pull/21110/files#r160463056

ice9js commented:
I updated the code and the description to provide some more insight in what we're trying to accomplish. The bigger changes I made are:

- I removed `requesting` state branch as it's no longer necessary.
- I converted the state tree into the shape proposed by @alexsanford .
- I added a separate action to reset a lock and removed `reset` property from the two existing actions.
- I updated the `blocked` reducer to assume that `ZONINATOR_REQUEST_LOCK_ERROR` means a zone is `blocked`, as that's the only thing that can cause an error on that endpoint right now besides using the wrong IDs or the server being down, in which case a 'Zone not found' page will be displayed anyway.

alexsanford commented on client/extensions/zoninator/state/locks/reducer.js:
```diff
@@ -0,0 +1,37 @@
+} );
+
+export const created = createReducer( 0, {
+	[ ZONINATOR_RESET_LOCK ]: ( state, { time } ) => time,
```
Is there a reason to use `{ time }` here, and `action` below, or can we make them consistent?

ice9js replied:
I did it so I didn't have to re-use `created` variable name, which is already used as the reducer's name and this prohibited by our style guide.

ockham replied:
I think the destructured one is actually quite elegant; in order to avoid variable name shadowing for the others, you could've aliased them, e.g.

```js
export const expires = createReducer( 0, {
	[ ZONINATOR_UPDATE_LOCK ]: ( state, { expires: expiresValue } ) => expiresValue,
} );
```
